File: _fable_pykg/src/z3.py
from __future__ import annotations
from _fable_pykg_infr import (solve_deps, TupleCons, create_tuple, ge_tuple, le_tuple, get_major, get_minor, get_micro)
from typing import (List, Optional, Tuple, Any)
from z3 import (ModelRef, AstRef, Or, simplify)
from ..fable_modules.fable_library.option import map
from ..fable_modules.fable_library.reflection import (TypeInfo, class_type)
from ..fable_modules.fable_library.string import (to_console, printf)

def expr_17() -> TypeInfo:
    return class_type("FablePykg.Z3.z3Model", None, z3Model)


class z3Model:
    def __init__(self, m: ModelRef) -> None:
        self.m = m
    

z3Model_reflection = expr_17

def z3model__ctor_D649C1C(m: ModelRef) -> z3Model:
    return z3Model(m)


def solve(a: List[AstRef[bool]]) -> Optional[z3Model]:
    def arrow_18(m: ModelRef, a: List[AstRef[bool]]=a) -> z3Model:
        return z3model__ctor_D649C1C(m)
    
    return map(arrow_18, solve_deps(a))


def const_ver(major: int, minor: int, micro: int) -> AstRef[Tuple[int, int, int]]:
    return TupleCons(major, minor, micro)


def test() -> None:
    v : AstRef[Tuple[int, int, int]] = create_tuple("m")
    def arrow_20(__unit: Any=None) -> AstRef[bool]:
        arg1_2 : AstRef[bool]
        arg1_4 : AstRef[Tuple[int, int, int]] = const_ver(1, 2, 9)
        arg1_2 = v == arg1_4
        def arrow_19(__unit: Any=None) -> AstRef[bool]:
            arg1_3 : AstRef[Tuple[int, int, int]] = const_ver(1, 2, 6)
            return v == arg1_3
        
        return Or(arrow_19(), arg1_2)
    
    model : Optional[z3Model] = solve([ge_tuple(v, const_ver(1, 2, 3)), le_tuple(v, const_ver(1, 2, 8)), arrow_20()])
    if model is not None:
        model_1 : z3Model = model
        tupled_arg : Tuple[int, int, int]
        x : AstRef[Tuple[int, int, int]] = model_1.m[v]
        def arrow_21(__unit: Any=None) -> int:
            a_2 : AstRef[int] = simplify(get_major(x))
            return a_2.as_long()
        
        def arrow_22(__unit: Any=None) -> int:
            a_4 : AstRef[int] = simplify(get_minor(x))
            return a_4.as_long()
        
        def arrow_23(__unit: Any=None) -> int:
            a_6 : AstRef[int] = simplify(get_micro(x))
            return a_6.as_long()
        
        tupled_arg = (arrow_21(), arrow_22(), arrow_23())
        to_console(printf("%A"))((tupled_arg[0], tupled_arg[1], tupled_arg[2]))
    
    else: 
        raise Exception("solution failed")
    

File: src/Z3.fs
module rec FablePykg.Z3
#nowarn "86"

open Fable.Core
type version = int * int * int


[<Import("AstRef", "z3")>]
type z3<'a> =

    [<Import("Or", "z3")>]
    static member (.||) (a: z3<bool>, b: z3<bool>): z3<bool> = nativeOnly

    [<Import("And", "z3")>]
    static member (.&&) (a: z3<bool>, b: z3<bool>): z3<bool> = nativeOnly


    [<Emit("$0 + $1")>]
    static member (+) (a: z3<int>, b: z3<int>) = nativeOnly

    [<Import("lt_tuple", "_fable_pykg_infr.z3_dep_solver")>]
    static member (.<) (a: z3<version>, b: z3<version>): z3<bool> = nativeOnly

    [<Import("le_tuple", "_fable_pykg_infr.z3_dep_solver")>]
    static member (.<=) (a: z3<version>, b: z3<version>): z3<bool> = nativeOnly

    [<Import("gt_tuple", "_fable_pykg_infr.z3_dep_solver")>]
    static member inline (.>) (a: z3<version>, b: z3<version>): z3<bool> =  nativeOnly

    [<Import("ge_tuple", "_fable_pykg_infr.z3_dep_solver")>]
    static member inline (.>=) (a: z3<version>, b: z3<version>): z3<bool> = b .<= a

    [<Emit("$0 == $1")>]
    static member (.=) (a: z3<version>, b: z3<version>): z3<bool> = nativeOnly

    [<Emit("$0 != $1")>]
    static member (.!=) (a: z3<version>, b: z3<version>): z3<bool> = nativeOnly
    // [<Import("eq_tuple", "_fable_pykg_infr.z3_dep_solver")>]
    // let (=) (a: z3<version>) (b: z3<version>) = nativeOnly

    // [<Import("gt_tuple", "_fable_pykg_infr.z3_dep_solver")>]
    // let (>) (a: z3<version>) (b: z3<version>) = nativeOnly

    // [<Import("ge_tuple", "_fable_pykg_infr.z3_dep_solver")>]
    // let (>=) (a: z3<version>) (b: z3<version>) = nativeOnly

    // [<Import("le_tuple", "_fable_pykg_infr.z3_dep_solver")>]
    // let (<=) (a: z3<version>) (b: z3<version>) = nativeOnly


[<Import("simplify", "z3")>]
let simplify (a: z3<'a>): z3<'a> = nativeOnly

[<Import("get_major", "_fable_pykg_infr.z3_dep_solver")>]
let get_major (a: z3<version>): z3<int> = nativeOnly

[<Import("get_minor", "_fable_pykg_infr.z3_dep_solver")>]
let get_minor (a: z3<version>): z3<int> = nativeOnly

[<Import("get_micro", "_fable_pykg_infr.z3_dep_solver")>]
let get_micro (a: z3<version>): z3<int> = nativeOnly

[<Emit("$0.as_long()")>]
let unwrapInt (a: z3<int>): int = nativeOnly


[<Import("ModelRef", "z3")>]
type _z3Model =
    [<Emit("$0[$1]")>]
    member self._get (a: z3<version>): z3<version> = nativeOnly


type z3Model( m: _z3Model ) =

    member inline _.Item with get (a: z3<version>): version =
        let x = m._get a
        let major = get_major x |> simplify |> unwrapInt
        let minor = get_minor x |> simplify |> unwrapInt
        let micro = get_micro x |> simplify |> unwrapInt
        (major, minor, micro)



[<Import("solve_deps", "_fable_pykg_infr.z3_dep_solver")>]
let _solve (a: z3<bool> array): option<_z3Model> = nativeOnly

let solve (a: z3<bool> array): option<z3Model> = Option.map (fun m -> z3Model m) (_solve a)

[<Import("If", "z3")>]
let If<'a> (a: z3<bool>) (a3: z3<'a>) (a4: z3<'a>) = nativeOnly

[<Import("TupleCons", "_fable_pykg_infr.z3_dep_solver")>]
let _constVersion (major: int) (minor: int) (micro: int): z3<version> = nativeOnly

[<Import("create_tuple", "_fable_pykg_infr.z3_dep_solver")>]
let varVer (name: string) = nativeOnly

let constVer (major: int, minor: int, micro: int): z3<version> =
    _constVersion major minor micro


let test () =

    let v = varVer "m"
    let model =
        solve [|
            v .>= constVer(1, 2, 3)
            v .<= constVer(1, 2, 8)
            (v .= constVer(1, 2, 6)) .|| (v .= constVer(1, 2, 9))
        |]
    match model with
    | None -> failwith "solution failed"
    | Some model ->
    let sol = model.[v]
    printf "%A" <| sol
